[sharded_tensor] easier initialization for Shard

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #66351
* #64481

This add the ability for use to just provide shard_offsets and optionally rank, to construct a local shard, instead of knowing there's a ShardedMetadata. Under the hood, we will construct the ShardedMetadata by inferring shard_lengths and device from the local tensor.

Differential Revision: [D31519919](https://our.internmc.facebook.com/intern/diff/D31519919/)

Commit:
[sharded_tensor] easier initialization for Shard

This add the ability for use to just provide shard_offsets and optionally rank, to construct a local shard, instead of knowing there's a ShardedMetadata. Under the hood, we will construct the ShardedMetadata by inferring shard_lengths and device from the local tensor.

Differential Revision: [D31519919](https://our.internmc.facebook.com/intern/diff/D31519919/)

[ghstack-poisoned]

pritamdamania87 reviewed (commented):
Could you also add a short description of the use case which needs this sort of construction?

bowangbj reviewed (commented):
Thanks Wanchao.

Commit:
Update on "[sharded_tensor] easier initialization for Shard"


This add the ability for use to just provide shard_offsets and optionally rank, to construct a local shard, instead of knowing there's a ShardedMetadata. Under the hood, we will construct the ShardedMetadata by inferring shard_lengths and device from the local tensor.

Differential Revision: [D31519919](https://our.internmc.facebook.com/intern/diff/D31519919/)

[ghstack-poisoned]

pritamdamania87 reviewed (changes requested):
@wanchaol Maybe your changes weren't pushed to gh successfully? I don't see any updates to the PR in the latest version.

Commit:
Update on "[sharded_tensor] easier initialization for Shard"


This add the ability for use to just provide shard_offsets and optionally rank, to construct a local shard, instead of knowing there's a ShardedMetadata. Under the hood, we will construct the ShardedMetadata by inferring shard_lengths and device from the local tensor.

Differential Revision: [D31519919](https://our.internmc.facebook.com/intern/diff/D31519919/)

[ghstack-poisoned]

Commit:
Update on "[sharded_tensor] easier initialization for Shard"


This add the ability for use to just provide shard_offsets and optionally rank, to construct a local shard, instead of knowing there's a ShardedMetadata. Under the hood, we will construct the ShardedMetadata by inferring shard_lengths and device from the local tensor.

Differential Revision: [D31519919](https://our.internmc.facebook.com/intern/diff/D31519919/)

[ghstack-poisoned]

Commit:
Update on "[sharded_tensor] easier initialization for Shard"


This add the ability for use to just provide shard_offsets and optionally rank, to construct a local shard, instead of knowing there's a ShardedMetadata. Under the hood, we will construct the ShardedMetadata by inferring shard_lengths and device from the local tensor.

Differential Revision: [D31519919](https://our.internmc.facebook.com/intern/diff/D31519919/)

[ghstack-poisoned]

Commit:
Update on "[sharded_tensor] easier initialization for Shard"


This add the ability for use to just provide shard_offsets and optionally rank, to construct a local shard, instead of knowing there's a ShardedMetadata. Under the hood, we will construct the ShardedMetadata by inferring shard_lengths and device from the local tensor.

Differential Revision: [D31519919](https://our.internmc.facebook.com/intern/diff/D31519919/)

[ghstack-poisoned]

Commit:
Update on "[sharded_tensor] easier initialization for Shard"


This add the ability for use to just provide shard_offsets and optionally rank, to construct a local shard, instead of knowing there's a ShardedMetadata. Under the hood, we will construct the ShardedMetadata by inferring shard_lengths and device from the local tensor.

Differential Revision: [D31519919](https://our.internmc.facebook.com/intern/diff/D31519919/)

[ghstack-poisoned]

Commit:
Update on "[sharded_tensor] easier initialization for Shard"


This add the ability for use to just provide shard_offsets and optionally rank, to construct a local shard, instead of knowing there's a ShardedMetadata. Under the hood, we will construct the ShardedMetadata by inferring shard_lengths and device from the local tensor.

Differential Revision: [D31519919](https://our.internmc.facebook.com/intern/diff/D31519919/)

[ghstack-poisoned]

Commit:
Update on "[sharded_tensor] easier initialization for Shard"


This add the ability for use to just provide shard_offsets and optionally rank, to construct a local shard, instead of knowing there's a ShardedMetadata. Under the hood, we will construct the ShardedMetadata by inferring shard_lengths and device from the local tensor.

Differential Revision: [D31519919](https://our.internmc.facebook.com/intern/diff/D31519919/)

[ghstack-poisoned]

Commit:
Update on "[sharded_tensor] easier initialization for Shard"


This add the ability for use to just provide shard_offsets and optionally rank, to construct a local shard, instead of knowing there's a ShardedMetadata. Under the hood, we will construct the ShardedMetadata by inferring shard_lengths and device from the local tensor.

Differential Revision: [D31519919](https://our.internmc.facebook.com/intern/diff/D31519919/)

[ghstack-poisoned]

Commit:
Update on "[sharded_tensor] easier initialization for Shard"


This add the ability for use to just provide shard_offsets and optionally rank, to construct a local shard, instead of knowing there's a ShardedMetadata. Under the hood, we will construct the ShardedMetadata by inferring shard_lengths and device from the local tensor.

Differential Revision: [D31519919](https://our.internmc.facebook.com/intern/diff/D31519919/)

[ghstack-poisoned]

Commit:
Update on "[sharded_tensor] easier initialization for Shard"


This add the ability for use to just provide shard_offsets and optionally rank, to construct a local shard, instead of knowing there's a ShardedMetadata. Under the hood, we will construct the ShardedMetadata by inferring shard_lengths and device from the local tensor.

Differential Revision: [D31519919](https://our.internmc.facebook.com/intern/diff/D31519919/)

[ghstack-poisoned]

Commit:
Update on "[sharded_tensor] easier initialization for Shard"


This add the ability for use to just provide shard_offsets and optionally rank, to construct a local shard, instead of knowing there's a ShardedMetadata. Under the hood, we will construct the ShardedMetadata by inferring shard_lengths and device from the local tensor.

Differential Revision: [D31519919](https://our.internmc.facebook.com/intern/diff/D31519919/)

[ghstack-poisoned]

Commit:
Update on "[sharded_tensor] easier initialization for Shard"


This add the ability for use to just provide shard_offsets and optionally rank, to construct a local shard, instead of knowing there's a ShardedMetadata. Under the hood, we will construct the ShardedMetadata by inferring shard_lengths and device from the local tensor.

Differential Revision: [D31519919](https://our.internmc.facebook.com/intern/diff/D31519919/)

[ghstack-poisoned]
